File: backend/app/routers/agent.py
"""Agent chat endpoint — LLM-powered assistant for list building and game queries."""

from __future__ import annotations

import logging
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from app.agent.agent import InfinityAgent
from app.agent.game_data.loader import GameDataLoader
from app.agent.schemas import ChatRequest, ChatResponse
from app.auth import get_current_user
from app.config import settings
from app.database import get_session
from app.models.user import User

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/agent", tags=["agent"])


def _get_provider():
    provider_name = settings.llm_provider.lower()
    if provider_name == "anthropic":
        if not settings.anthropic_api_key:
            raise HTTPException(
                status_code=503,
                detail="AI features are not configured (missing API key).",
            )
        from app.agent.providers.anthropic import AnthropicProvider

        return AnthropicProvider(
            api_key=settings.anthropic_api_key,
            model=settings.llm_model,
        )
    elif provider_name == "gemini":
        from app.agent.providers.gemini import GeminiProvider

        return GeminiProvider()
    else:
        raise HTTPException(
            status_code=503,
            detail=f"Unknown LLM provider: {provider_name}",
        )


async def _check_and_increment_usage(user_id: str, db: AsyncSession) -> None:
    year_month = datetime.now().strftime("%Y-%m")
    result = await db.execute(
        text(
            """
            INSERT INTO ai_usage (user_id, year_month, message_count)
            VALUES (:user_id, :year_month, 1)
            ON CONFLICT (user_id, year_month)
            DO UPDATE SET message_count = ai_usage.message_count + 1,
                          last_used = NOW()
            RETURNING message_count
            """
        ),
        {"user_id": user_id, "year_month": year_month},
    )
    await db.commit()
    count = result.scalar_one()
    if count > settings.ai_monthly_limit:
        raise HTTPException(
            status_code=429,
            detail=(
                f"Monthly AI message limit reached ({settings.ai_monthly_limit} messages). "
                "Limit resets at the start of next month."
            ),
        )


@router.post("/chat")
async def chat(
    request: ChatRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_session),
) -> ChatResponse:
    """Chat with the Infinity AI assistant. Requires authentication."""
    await _check_and_increment_usage(user.id, db)

    loader = GameDataLoader.get_instance()
    provider = _get_provider()
    agent = InfinityAgent(provider=provider, loader=loader)

    try:
        response = await agent.chat(
            message=request.message,
            history=request.history,
            context=request.context,
            db=db,
        )
    except Exception as e:
        logger.exception("Agent chat failed for user %s: %s", user.id, e)
        raise HTTPException(
            status_code=500, detail="AI assistant encountered an error."
        ) from e

    return ChatResponse(
        reply=response.text,
        tools_used=response.tool_calls_made,
        input_tokens=response.input_tokens,
        output_tokens=response.output_tokens,
    )

File: backend-ts/src/routes/agent.ts
import { OpenAPIHono, createRoute, z } from '@hono/zod-openapi';
import { config } from '../config.js';
import { requireUser, type AuthVars } from '../auth/middleware.js';
import { InfinityAgent, type ChatContext, type ChatMessage } from '../agent/agent.js';
import { AnthropicProvider } from '../agent/providers/anthropic.js';
import { GeminiProvider } from '../agent/providers/gemini.js';
import { GameDataLoader } from '../agent/gameData/loader.js';
import type { LLMProvider } from '../agent/providers/base.js';
import { checkAndIncrementUsage, RateLimitError } from '../lib/usage.js';

const router = new OpenAPIHono<{ Variables: AuthVars }>();

router.use('*', requireUser);

const ListContextUnitSchema = z.object({
    name: z.string(),
    isc: z.string(),
    loadout: z.string().optional(),
    points: z.number().optional(),
    swc: z.number().optional(),
}).openapi('ListContextUnit');

const ChatContextSchema = z.object({
    faction_id: z.number().int().nullable().optional(),
    faction_name: z.string().nullable().optional(),
    list_points: z.number().nullable().optional(),
    list_swc: z.number().nullable().optional(),
    list_units: z.array(ListContextUnitSchema).optional(),
}).openapi('ChatContext');

const ChatMessageSchema = z.object({
    role: z.enum(['user', 'assistant']),
    content: z.string(),
}).openapi('ChatMessage');

const ChatRequestSchema = z.object({
    message: z.string().max(2000),
    history: z.array(ChatMessageSchema).max(20).optional(),
    context: ChatContextSchema.optional(),
}).openapi('ChatRequest');

const ChatResponseSchema = z.object({
    reply: z.string(),
    tools_used: z.array(z.string()),
    input_tokens: z.number().int(),
    output_tokens: z.number().int(),
}).openapi('ChatResponse');

const ErrorSchema = z.object({
    detail: z.union([z.string(), z.array(z.unknown())]),
}).openapi('Error');

class ProviderUnavailable extends Error {}

function getProvider(): LLMProvider {
    if (config.llmProvider === 'anthropic') {
        if (!config.anthropicApiKey) {
            throw new ProviderUnavailable('AI features are not configured (missing API key).');
        }
        return new AnthropicProvider(config.anthropicApiKey, config.llmModel);
    }
    if (config.llmProvider === 'gemini') return new GeminiProvider();
    throw new ProviderUnavailable(`Unknown LLM provider: ${config.llmProvider}`);
}

router.openapi(
    createRoute({
        method: 'post',
        path: '/chat',
        tags: ['agent'],
        summary: 'Chat with the Infinity AI agent',
        request: {
            body: { content: { 'application/json': { schema: ChatRequestSchema } }, required: true },
        },
        responses: {
            200: { content: { 'application/json': { schema: ChatResponseSchema } }, description: 'Reply' },
            401: { content: { 'application/json': { schema: ErrorSchema } }, description: 'Unauthorized' },
            422: { content: { 'application/json': { schema: ErrorSchema } }, description: 'Validation error' },
            429: { content: { 'application/json': { schema: ErrorSchema } }, description: 'Rate limited' },
            500: { content: { 'application/json': { schema: ErrorSchema } }, description: 'Server error' },
            503: { content: { 'application/json': { schema: ErrorSchema } }, description: 'AI provider unavailable' },
        },
    }),
    async c => {
        const user = c.get('user');
        const data = c.req.valid('json');

        try {
            await checkAndIncrementUsage(user.id, config.aiMonthlyLimit);
        } catch (e) {
            if (e instanceof RateLimitError) return c.json({ detail: e.message }, 429);
            throw e;
        }

        let provider: LLMProvider;
        try {
            provider = getProvider();
        } catch (e) {
            if (e instanceof ProviderUnavailable) return c.json({ detail: e.message }, 503);
            throw e;
        }

        const loader = GameDataLoader.getInstance();
        const agent = new InfinityAgent(provider, loader);

        const history = (data.history ?? []) as ChatMessage[];
        const context = (data.context ?? {}) as ChatContext;

        try {
            const response = await agent.chat(data.message, history, context);
            return c.json({
                reply: response.text,
                tools_used: response.toolCallsMade,
                input_tokens: response.inputTokens,
                output_tokens: response.outputTokens,
            }, 200);
        } catch (e) {
            console.error('Agent chat failed for user', user.id, e);
            return c.json({ detail: 'AI assistant encountered an error.' }, 500);
        }
    },
);

export default router;
